src/lambda/commands/createNewSamApp.ts

```diff
@@ -201,18 +203,52 @@ export async function createNewSamApplication(
             true
         )
 
-        const newLaunchConfigs = await addInitialLaunchConfiguration(
-            extContext,
-            vscode.workspace.getWorkspaceFolder(uri)!,
-            uri
-        )
-        await vscode.window.showTextDocument(uri)
-        if (newLaunchConfigs && newLaunchConfigs.length > 0) {
-            showCompletionNotification(config.name, `"${newLaunchConfigs.map(config => config.name).join('", "')}"`)
-        }
-        activationLaunchPath.clearLaunchPath()
+        // Race condition where SAM app is created but template doesn't register in time.
+        // Poll for 5 seconds, otherwise direct user to codelens.
+        const isTemplateRegistered = await waitUntil(async () => {
+            return ext.templateRegistry.getRegisteredTemplate(uri.fsPath)
+        })
+        try {
+            if (isTemplateRegistered) {
+                const newLaunchConfigs = await addInitialLaunchConfiguration(
+                    extContext,
+                    vscode.workspace.getWorkspaceFolder(uri)!,
+                    uri
+                )
+                if (newLaunchConfigs && newLaunchConfigs.length > 0) {
+                    showCompletionNotification(
+                        config.name,
+                        `"${newLaunchConfigs.map(config => config.name).join('", "')}"`
+                    )
+                }
+                reason = 'complete'
+            } else {
+                throw new Error()
```

Comment:
could avoid using exceptions for control-flow, by factoring out the `catch` block into a function. Actually, I don't think the catch block is necessary: there is already a try-catch surrounding this to handle exceptions. So we don't need another one here. We only need to handle the "failed to generate" case, in this `else` block.

The `finally` block will need to be duplicated. Or we could declare some local functions to avoid duplication.

Reply:
`addInitialLaunchConfiguration` already shows an error if it fails. AFAIK `addInitialLaunchConfiguration` should only fail if there's a filesystem issue, unrelated to the purpose of this PR. If that's correct then `addInitialLaunchConfiguration` doesn't need to throw and we don't need error-control-flow here.

Reply:
How about this:
* have the `addInitialLaunchConfiguration` message also provide a link to the docs to create a launch config
* remove error-based control flow
* present the same message that this PR presents in the else (instead of the error)

I'm not convinced that `addInitialLaunchConfiguration` only fails on a filesystem issue; it uses VS Code-specific functionality to add the launch config, and it also fails on the `launch.json` being dirty (honestly the main reason why I see the error pop up). At the very least, it's worth presenting a way from that error dialog to create a launch config manually.

Reply:
ok